s3datarepository/iam.go

```diff
@@ -241,132 +422,292 @@ func (s *S3Repository) GrantAccess(ctx context.Context, id string, derivative bo
+
+			// find out what role(s) correspond to this instance profile and what policies are attached to them
+			var ipOut *iam.GetInstanceProfileOutput
+			ipOut, err = s.IAM.GetInstanceProfileWithContext(ctx, &iam.GetInstanceProfileInput{
+				InstanceProfileName: aws.String(currentInstanceProfileName),
+			})
+			if err != nil {
+				return nil, ErrCode("failed to get information about current instance profile "+currentInstanceProfileName, err)
 			}
-			return nil
-		}()
-	})
 
-	// add role to instance profile
-	_, err = s.IAM.AddRoleToInstanceProfileWithContext(ctx, &iam.AddRoleToInstanceProfileInput{
-		InstanceProfileName: aws.String(roleName),
-		RoleName:            aws.String(roleName),
-	})
-	if err != nil {
-		return nil, ErrCode("failed to add role to instance profile "+roleName, err)
+			// TODO: we are _not_ considering role inline policies at this point, should we?
+			var currentPoliciesArn []string
+			for _, r := range ipOut.InstanceProfile.Roles {
+				log.Debugf("listing attached policies for role %s", aws.StringValue(r.RoleName))
+
+				var attachedRolePoliciesOut *iam.ListAttachedRolePoliciesOutput
+				attachedRolePoliciesOut, err = s.IAM.ListAttachedRolePoliciesWithContext(ctx, &iam.ListAttachedRolePoliciesInput{
+					RoleName: r.RoleName,
+				})
+				if err != nil {
+					return nil, ErrCode("failed to list attached policies for role "+aws.StringValue(r.RoleName), err)
+				}
+
+				if attachedRolePoliciesOut.AttachedPolicies == nil {
+					log.Warnf("no attached policies found for current role %s, there may be inline policies", aws.StringValue(r.RoleName))
+				}
+
+				for _, p := range attachedRolePoliciesOut.AttachedPolicies {
+					currentPoliciesArn = append(currentPoliciesArn, aws.StringValue(p.PolicyArn))
+				}
+			}
+
+			log.Infof("policies attached to the current instance profile: %s", currentPoliciesArn)
+
+			// attach current policies to our new role
+			for _, p := range currentPoliciesArn {
+				log.Debugf("attaching pre-existing policy %s to role %s", p, roleName)
+
+				_, err = s.IAM.AttachRolePolicyWithContext(ctx, &iam.AttachRolePolicyInput{
+					PolicyArn: aws.String(p),
+					RoleName:  aws.String(roleName),
+				})
+				if err != nil {
+					return nil, ErrCode("failed to attach policy "+p+" to role "+roleName, err)
+				}
+			}
+
+			// append policy detach from role to rollback tasks
+			rollBackTasks = append(rollBackTasks, func() error {
+				return func() error {
+					for _, p := range currentPoliciesArn {
+						log.Debugf("DetachRolePolicyWithContext: %s (%s)", p, roleName)
+						err = retry(3, 3*time.Second, func() error {
+							_, err = s.IAM.DetachRolePolicyWithContext(ctx, &iam.DetachRolePolicyInput{
+								PolicyArn: aws.String(p),
+								RoleName:  aws.String(roleName),
+							})
+							if err != nil {
+								log.Debugf("retrying, got error: %s", err)
+								return err
+							}
+							return nil
+						})
+						if err != nil {
+							return err
+						}
```

Comment:
do you want to return or try the next one?

Reply:
good point!